syncer/syncer.go

```diff
@@ -91,16 +94,18 @@ type Syncer struct {
 	awaitSyncedCh []chan struct{}
 	awaitSyncedMu sync.Mutex
 
-	shutdownCtx context.Context
-	cancelFunc  context.CancelFunc
+	// tomb is used to manage goroutines for graceful shutdown.
+	// note that tomb kills itself after its last goroutine exits. syncer's loop is the long-living goroutine
+	// that keeps tomb alive before the node shuts down.
+	// see https://github.com/go-tomb/tomb/issues/28
+	tomb tomb.Tomb
```

Comment:
i think https://pkg.go.dev/golang.org/x/sync/errgroup doesn't have this issue. very popular and it also has a simpler API

Reply:
@dshulyak it looks promising! i am quite annoyed with this issue.

so during shutdown, i call 
```
g.Go(func() error {
    return errShuttingDown
})
g.Wait()
```
did i read that correctly?
if so, the API is indeed simpler.

Reply:
it is usually used with context or another termination mechanism (a channel), so the code looks like:

```go
ctx, cancel := context.WIthCancel(context.Background())
g.Go(func() error {
    <-ctx.Done()
    return ctx.Err()
})
g.Wait()
```

But there is another detail that i missed. With errgroup, if Go called after ctx is canceled and Wait is called it will panic. and with tomb it will not try to spawn a goroutine and just exit, right? so the tomb is probably better for this use case, unless we want to make a small wrapper around errgroup

Reply:
I actually think this is not how Tomb was intended to be used. The way it should work is that you instantiate a tomb in main and then start all modules with it. Then it makes perfect sense that the tomb kills itself when the last goroutine dies.

To implement this concept incrementally, I would start by instantiating the tomb in main, but only use it to start the sync module. This way, the goroutine that runs the sync "event loop" will keep the tomb alive and you can then start other goroutines from that event loop.

I didn't know about `errgroup` before, which seems to perform a similar function, putting more of the heavy lifting on the context itself. It's more recently updated and more widely used, making me thing we should consider using it in place of tomb. I think it can pretty easily replace tomb in the code - unless either of you sees any issue with it, I would give it a try and see if it doesn't introduce new issues.

Reply:
I don't think it should be the responsibility of the goroutine manager to "prevent spawning new goroutines". The shutdown procedure might require spawning processes and that's fine. This should be our responsibility and the tool provides us a way to know if we're shutting down so WE can decide not to spawn. In Tomb you check if the `dying` channel is closed and in errgroup you check the `context.Done()` channel.

Reply:
i agree. but in order not to repeat ourselves (with selects on ctx.Done() before calling Go) we can write a small wrapper around errgroup that will do it always for us.

also i think it is not as simple as a single select, select on ctx.Done() and calling Go should be atomic, otherwise there will be a race with Wait. so some locking will be required.

Reply:
I don't think we should necessarily check `ctx.Done()` before starting goroutines.

As I see it, every `select` statement we put in the code should include a read from `ctx.Done()` and the context should be passed to any long-running method that can be canceled.

You want to be able to do shutdown procedures in parallel (e.g. flushing state to disk).

Reply:
I briefly considered implementing our own `select`, but it [looks like](https://stackoverflow.com/a/19992525/1225467) it would be a syntactic nightmare...

Reply:
the problem is that if you call Go after Wait is called errgroup will panic (the same as regular sync.WaitGroup). what you described regarding select is obviously needed but not enough if we will use the same group to manage goroutines that are spawned by other goroutines

Reply:
A WaitGroup will only panic if a positive `Add()` happens while there's a pending `Wait()` _and the counter is zero_. ([see docs](https://pkg.go.dev/sync#WaitGroup))

I don't know enough about errgroup, but if the problem is that it uses an underlying WaitGroup then it shouldn't be an issue: as long as there are running goroutines, more goroutines can be started (since the counter isn't zero) and once all goroutines have completed, there's no risk of some goroutine spawning another goroutine. Right?

Reply:
i believe there is an issue. see this code for an example:

```go
var g errgroup.Group
g.Go(func() {
    g.Go(func() {}) 
})

g.Wait() // you cannot guarantee that both goroutines will exit
```

Reply:
this is from the source code of Add

```
       if w != 0 && delta > 0 && v == int32(delta) {
		panic("sync: WaitGroup misuse: Add called concurrently with Wait")
	}
	if v > 0 || w == 0 {
		return
	}
	// This goroutine has set counter to 0 when waiters > 0.
	// Now there can't be concurrent mutations of state:
	// - Adds must not happen concurrently with Wait,
	// - Wait does not increment waiters if it sees counter == 0.
	// Still do a cheap sanity check to detect WaitGroup misuse.
	if *statep != state {
		panic("sync: WaitGroup misuse: Add called concurrently with Wait")
	}
```